docs/design/proposals/cephfs-snapshot-shallow-ro-vol.md

```diff
@@ -0,0 +1,274 @@
+
+To enable creating shallow volumes from snapshots that were provisioned by
+older versions of cephfs-csi (i.e. before this feature is introduced),
+`CreateVolume` for shallow volumes would also create the _refcount_ object in
+case it's missing. It would be initialized to two: the source snapshot and the
+newly created shallow volume.
+
+##### Concurrent access to _refcount_ objects
+
+We can't do operations on integers in RADOS objects atomically, which means we
+need to do all synchronization ourselves when modifying the value of the counter.
+
+From a single master perspective, concurrent access to _refcount_ can be
+protected by mutex locks and is easy to implement. `cephfs.ControllerServer`
+already offers `VolumeLocks`, `SnapshotLocks` and `OperationLocks`.
+
+Mutexes won't cover situations where multiple masters may be active though
+(during/after leader election), and modifying _refcount_ would still race in
+this case. Misbehaving master that lost leadership would still be able to
+access resources on the Ceph cluster. Racing on _refcount_ could then cause
+leaking the _refcount_ object itself and/or the subvolume snapshot, or the
+provisioned shallow volume could point to a subvolume snapshot that's already
+been deleted. This is not easily fixable without the individual master replicas
+being aware of the leader election process and is out-of-scope for this
+proposal.
+
+#### `CreateVolume`
+
+A read-only volume with snapshot source would be created under these conditions:
+
+1. `CreateVolumeRequest.volume_content_source` is a snapshot,
+1. `CreateVolumeRequest.volume_capabilities[*].access_mode` is any of read-only
+   volume access modes.
+1. Possibly other volume parameters in `CreateVolumeRequest.parameters`
+   specific to shallow volumes.
+
+`CreateVolumeResponse.Volume.volume_context` would then contain necessary
+information to identify the source subvolume / snapshot.
+
+Things to look out for:
+
+* _What's the volume size?_
+
+  It doesn't consume any space on the filesystem. `Volume.capacity_bytes` is
+  allowed to contain zero. We could use that.
+* _What should be the requested size when creating the volume (specified e.g.
+  in PVC)?_
+
+  This one is tricky. CSI spec allows for
+  `CreateVolumeRequest.capacity_range.{required_bytes,limit_bytes}` to be
+  zero. On the other hand,
+  `PersistentVolumeClaim.spec.resources.requests.storage` must be bigger
+  than zero. cephfs-csi doesn't care about the requested size (the volume
+  will be read-only, so it has no usable capacity) and would always set it
+  to zero. This shouldn't case any problems for the time being, but still
+  is something we should keep in mind.
+
+### `DeleteVolume`
+
+Volume deletion is trivial.
+
+### `CreateSnapshot`
+
+Snapshotting read-only volumes doesn't make sense in general, and should
+be rejected.
+
+### `ControllerExpandVolume`
+
+Same thing as above. Expanding read-only volumes doesn't make sense in general,
+and should be rejected.
```

Comment:
from the kubernetes point of view, it's a new PVC. consider adding CreateVolume from snapshot and CreateVolume from volume case here as well.